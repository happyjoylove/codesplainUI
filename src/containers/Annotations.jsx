import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { CardText } from 'material-ui/Card';
import { withRouter } from 'react-router';

import {
  closeAnnotationPanel,
  openAnnotationPanel,
} from '../actions/annotation';
import {
  saveAnnotation,
  saveExisting,
} from '../actions/app';
import AnnotationPanel from '../components/AnnotationPanel';
import {
  getAnnotatedLines,
  getNextAnnotation,
  getPreviousAnnotation,
  hasNextAnnotation,
  hasPreviousAnnotation,
} from '../util/annotations';
import CustomPropTypes from '../util/custom-prop-types';

export class Annotations extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      displayStatus: 'none',
      hasPreceedingAnnotation: false,
      hasProceedingAnnotation: false,
    };
    this.getNextAnnotation = this.getNextAnnotation.bind(this);
    this.getPreviousAnnotation = this.getPreviousAnnotation.bind(this);
    this.handleSaveAnnotation = this.handleSaveAnnotation.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    const {
      annotations,
      lineAnnotated: {
        lineNumber,
      },
    } = nextProps;
    const nextAnnotatedLines = getAnnotatedLines(annotations);
    this.setState({
      hasProceedingAnnotation: hasNextAnnotation(nextAnnotatedLines, lineNumber),
      hasPreceedingAnnotation: hasPreviousAnnotation(nextAnnotatedLines, lineNumber),
    });
  }

  getPreviousAnnotation() {
    const {
      lineAnnotated: {
        lineNumber: currentLineNumber,
      },
      annotations,
      displayAnnotationPanel,
    } = this.props;

    // Get the annotated lines
    const previous = getPreviousAnnotation(annotations, currentLineNumber);
    if (!previous) {
      return;
    }
    const toDisplay = {
      lineNumber: previous.lineNumber,
      lineText: previous.lineText,
    };
    displayAnnotationPanel(toDisplay);
  }

  getNextAnnotation() {
    const {
      lineAnnotated: {
        lineNumber: currentLineNumber,
      },
      annotations,
      displayAnnotationPanel,
    } = this.props;

    // Get the annotated lines
    const next = getNextAnnotation(annotations, currentLineNumber);
    if (!next) {
      return;
    }
    const toDisplay = {
      lineNumber: next.lineNumber,
      lineText: next.lineText,
    };
    displayAnnotationPanel(toDisplay);
  }

  handleSaveAnnotation(annotation) {
    const {
      lineAnnotated,
      saveAnnotationToState,
      saveExistingSnippet,
      snippetKey,
    } = this.props;

    const annotationData = {
      annotation,
      ...lineAnnotated,
    };

    saveAnnotationToState(annotationData);
    // Save the snippet only if this snippet has already been saved.
    if (snippetKey) {
      saveExistingSnippet();
    }
  }

  render() {
    const {
      annotation,
      hideAnnotationPanel,
      isDisplayingAnnotation,
      lineAnnotated,
      readOnly,
    } = this.props;

    const {
      hasPreceedingAnnotation,
      hasProceedingAnnotation,
    } = this.state;

    if (!isDisplayingAnnotation) {
      const prompt = readOnly ?
        'Click on a line number to add an annotation or display one' :
        'Lock this snippet to add annotations';
      return (
        <CardText>{prompt}</CardText>
      );
    }
    return (
      <AnnotationPanel
        annotation={annotation}
        closeAnnotation={hideAnnotationPanel}
        getNextAnnotation={this.getNextAnnotation}
        getPreviousAnnotation={this.getPreviousAnnotation}
        hasNextAnnotation={hasProceedingAnnotation}
        hasPrevAnnotation={hasPreceedingAnnotation}
        lineAnnotated={lineAnnotated}
        saveAnnotation={this.handleSaveAnnotation}
      />
    );
  }
}

Annotations.propTypes = {
  annotation: PropTypes.string.isRequired,
  annotations: CustomPropTypes.annotations.isRequired,
  displayAnnotationPanel: PropTypes.func.isRequired,
  hideAnnotationPanel: PropTypes.func.isRequired,
  isDisplayingAnnotation: PropTypes.bool.isRequired,
  lineAnnotated: CustomPropTypes.lineAnnotated.isRequired,
  readOnly: PropTypes.bool.isRequired,
  saveAnnotationToState: PropTypes.func.isRequired,
  saveExistingSnippet: PropTypes.func.isRequired,
  snippetKey: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => {
  const {
    annotation: {
      isDisplayingAnnotation,
      lineAnnotated,
    },
    app: {
      annotations,
      readOnly,
      snippetKey,
    },
  } = state;
  const { lineNumber } = lineAnnotated;
  // Check that
  const annotation = (isDisplayingAnnotation && annotations[lineNumber] && annotations[lineNumber].annotation) || '';
  return {
    annotation,
    annotations,
    isDisplayingAnnotation,
    lineAnnotated,
    readOnly,
    snippetKey,
  };
};

export const mapDispatchToProps = dispatch => ({
  displayAnnotationPanel: (toDisplay) => {
    dispatch(openAnnotationPanel(toDisplay));
  },
  hideAnnotationPanel: () => {
    dispatch(closeAnnotationPanel());
  },
  saveAnnotationToState: (annotation) => {
    dispatch(saveAnnotation(annotation));
  },
  saveExistingSnippet: () => {
    dispatch(saveExisting());
  },
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Annotations));
