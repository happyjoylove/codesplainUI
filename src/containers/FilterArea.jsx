import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import {
  Card,
  CardText,
  CardTitle,
} from 'material-ui';

import {
  resetFilters,
  selectAllFilters,
  setRuleFilters,
} from '../actions/app';
import FilterAreaActions from '../components/FilterAreaActions';
import RulesSelector from '../components/RulesSelector';
import CustomPropTypes from '../util/custom-prop-types';

const filterCount = filters => _.keys(filters).length;

export class FilterArea extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isExpanded: filterCount(props.filters) !== 0,
    };
    this.handleRuleSelected = this.handleRuleSelected.bind(this);
    this.handleClearAllFilters = this.handleClearAllFilters.bind(this);
    this.handleSelectAllFilters = this.handleSelectAllFilters.bind(this);
    this.handleExpandChange = this.handleExpandChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    const numberOfFilters = filterCount(this.props.filters);
    const numberOfNextFilters = filterCount(nextProps.filters);
    // Do nothing if the number of filters hasn't changed
    if (numberOfFilters === numberOfNextFilters) {
      return;
    }
    // This component's isExpanded property is based on whether there are filters
    // to display or not. If there aren't filters to display, close the panel;
    // if there are, it should be expanded.
    this.setState({
      isExpanded: Boolean(numberOfNextFilters),
    });
  }

  handleRuleSelected(filterName) {
    const { dispatch, filters } = this.props;
    const newFilters = _.cloneDeep(filters);
    newFilters[filterName].selected = !newFilters[filterName].selected;
    dispatch(setRuleFilters(newFilters));
  }

  handleExpandChange(isExpanded) {
    this.setState({ isExpanded });
  }

  handleClearAllFilters() {
    const { dispatch } = this.props;
    dispatch(resetFilters());
  }

  handleSelectAllFilters() {
    const { dispatch } = this.props;
    dispatch(selectAllFilters());
  }

  render() {
    const { filters } = this.props;
    const { isExpanded } = this.state;
    const doesHaveFilters = Boolean(filterCount(filters));
    const filterAreaActions = doesHaveFilters && isExpanded ?
      (<FilterAreaActions
        clearAll={this.handleClearAllFilters}
        selectAll={this.handleSelectAllFilters}
      />) : null;
    return (
      <Card
        expanded={isExpanded}
        onExpandChange={this.handleExpandChange}
      >
        <CardTitle
          actAsExpander={doesHaveFilters}
          title={<h2>Filters</h2>}
          showExpandableButton={doesHaveFilters}
        />
        <CardText
          expandable
        >
          <RulesSelector
            filters={filters}
            onRuleSelected={this.handleRuleSelected}
          />
        </CardText>
        { filterAreaActions }
      </Card>
    );
  }
}

FilterArea.propTypes = {
  filters: CustomPropTypes.filters.isRequired,
};

const mapStateToProps = state => ({
  filters: state.app.filters,
});

export default connect(mapStateToProps)(FilterArea);
