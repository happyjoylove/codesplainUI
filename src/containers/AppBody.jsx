import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import cookie from 'react-cookie';

import Annotations from './Annotations';
import FilterArea from './FilterArea';
import SnippetArea from './SnippetArea';
import ReferenceArea from '../components/ReferenceArea';
import {
  loadSnippet,
  restoreState,
} from '../actions/app';
import { setPermissions } from '../actions/permissions';
import { restoreUserCredentials } from '../actions/user';
import { removeDeprecatedFiltersFromState } from '../util/codemirror-utils';
import { setDefaults } from '../util/state-management';

export class AppBody extends Component {
  componentDidMount() {
    const {
      dispatch,
      router,
    } = this.props;
    const {
      id: snippetKey,
      username,
    } = router.params;

    if (!username && !snippetKey) {
      // This is a new snippet for the current user, enable all permissions
      const permissions = {
        canRead: true,
        canEdit: true,
      };
      dispatch(setPermissions(permissions));

      // Reload the state if returning from login.
      const signInState = cookie.load('signInState');
      if (signInState) {
        cookie.remove('signInState', { path: '/' });
        cookie.remove('signInRedirect', { path: '/' });
        dispatch(restoreState(signInState));
      }
      return;
    }

    // Restore the user's credentials into state
    dispatch(restoreUserCredentials(cookie.load('token'), cookie.load('username')));

    dispatch(loadSnippet(username, snippetKey))
      .then((res) => {
        // Normalize the app state received from S3
        const appState = setDefaults(removeDeprecatedFiltersFromState(res.data));
        // Snippet may not have a S3 key value saved; set it to the URL's id
        // param if the state object is lacking it
        if (!appState.snippetKey) {
          appState.snippetKey = snippetKey;
        }
        // Restore the application's state
        dispatch(restoreState(appState));

        const permissions = {
          canRead: true,
          // Currently, users may only edit a file they own
          canEdit: (username === cookie.load('username')),
        };
        dispatch(setPermissions(permissions));

        // Reroute if not at the 'correct' location
        // So /:username/snippets/:id -> /:username/:id
        const nextRoute = `/${username}/${encodeURIComponent(snippetKey)}`;
        if (router.location.pathname !== nextRoute) {
          router.push(nextRoute);
        }
      }, () => {
        // Failed to get the snippet, either bad URL or unauthorized
        router.push('/');
      });
  }

  render() {
    return (
      <div style={{ height: '85%' }}>
        <div
          style={{
            display: 'flex',
            flexFlow: 'row wrap',
            height: '100%',
          }}
        >
          <SnippetArea />
          <div
            style={{
              display: 'flex',
              flexFlow: 'column wrap',
              height: '100%',
              flex: '1 0',
              minWidth: '300px',
            }}
          >
            <FilterArea />
            <Annotations />
          </div>
        </div>
        <div>
          <ReferenceArea />
        </div>
      </div>
    );
  }
}

export default withRouter(connect()(AppBody));
