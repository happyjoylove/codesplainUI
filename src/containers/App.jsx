import React from 'react';

import AppBody from './AppBody';
import CodesplainAppBar from './CodesplainAppBar';
import NotificationSnackbar from '../components/NotificationSnackbar';
import ReferenceArea from '../components/ReferenceArea';

const styles = {
  container: {
    height: 'inherit',
    display: 'flex',
    flexFlow: 'column nowrap',
    justifyContent: 'flex-start',
    overflow: 'auto',
  },
};

export const App = () => (
  <div style={styles.container}>
    <CodesplainAppBar />
    <AppBody />
    <ReferenceArea />
    <NotificationSnackbar />
  </div>
);

export default App;
