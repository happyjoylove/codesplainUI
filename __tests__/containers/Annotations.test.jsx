import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import { Annotations, mapDispatchToProps } from '../../src/containers/Annotations';

const defaultProps = {
  annotation: '',
  annotations: {},
  displayAnnotationPanel: jest.fn(),
  hideAnnotationPanel: jest.fn(),
  isDisplayingAnnotation: false,
  lineAnnotated: {},
  readOnly: false,
  saveAnnotationToState: jest.fn(),
  saveExistingSnippet: jest.fn(),
  snippetKey: '',
};

const muiTheme = getMuiTheme();
const shallowWithContext = node => shallow(node, { context: { muiTheme } });

const setup = (props = {}) => (
  shallowWithContext(<Annotations {...defaultProps} {...props} />)
);

describe('<Annotations />', () => {
  describe('snapshots', () => {
    it('matches when snippet is not locked', () => {
      const wrapper = setup();
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
    it('matches when snippet is locked and no annotation is displayed', () => {
      const wrapper = setup({ readOnly: true });
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
    it('matches when snippet is locked and an annotation is displayed', () => {
      const props = {
        annotation: 'annotation',
        isDisplayingAnnotation: true,
        lineAnnotated: { lineNumber: 1, lineText: 'line of code' },
        readOnly: true,
      };
      const wrapper = setup(props);
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('mapDispatchToProps', () => {
    it('matches snapshot', () => {
      expect(mapDispatchToProps).toMatchSnapshot();
    });
  });
});
