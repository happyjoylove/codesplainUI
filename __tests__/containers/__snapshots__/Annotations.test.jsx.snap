exports[`<Annotations /> mapDispatchToProps matches snapshot 1`] = `[Function]`;

exports[`<Annotations /> snapshots matches when snippet is locked and an annotation is displayed 1`] = `
<AnnotationPanel
  annotation="annotation"
  closeAnnotation={[Function]}
  getNextAnnotation={[Function]}
  getPreviousAnnotation={[Function]}
  hasNextAnnotation={false}
  hasPrevAnnotation={false}
  lineAnnotated={
    Object {
      "lineNumber": 1,
      "lineText": "line of code",
    }
  }
  saveAnnotation={[Function]} />
`;

exports[`<Annotations /> snapshots matches when snippet is locked and no annotation is displayed 1`] = `
<CardText>
  Click on a line number to add an annotation or display one
</CardText>
`;

exports[`<Annotations /> snapshots matches when snippet is not locked 1`] = `
<CardText>
  Lock this snippet to add annotations
</CardText>
`;
